Test/Tests/Network/DNS.Tests.ps1

```diff
@@ -0,0 +1,395 @@
+    $IP = $ContainerNetInfo.IPAddress
+    Write-Log "IP of $ContainerID : $IP"
+
+    return $IP
+}
+
+# This function is used only to generate command
+# that will be passed to docker exec.
+# $Hostname will be substituted.
+function Resolve-DNSLocally {
+    $resolved = (Resolve-DnsName -Name $Hostname -Type A -ErrorAction SilentlyContinue)
+
+    if($error.Count -eq 0) {
+        Write-Host "found"
+        $resolved[0].IPAddress
+    } else {
+        Write-Host "error"
+        $error[0].CategoryInfo.Category
+    }
+}
+$ResolveDNSLocallyCommand = (${function:Resolve-DNSLocally} -replace "`n|`r", ";")
+
+function Resolve-DNS {
+    Param (
+        [Parameter(Mandatory=$true)] [PSSessionT] $Session,
+        [Parameter(Mandatory=$true)] [String] $ContainerName,
+        [Parameter(Mandatory=$true)] [String] $Hostname
+    )
+
+    $Command = $ResolveDNSLocallyCommand -replace '\$Hostname', ('"'+$Hostname+'"')
+
+    $Result = (Invoke-Command -Session $Session -ScriptBlock {
+        docker exec $Using:ContainerName powershell $Using:Command
+    }).Split([Environment]::NewLine)
+
+    Write-Log "Resolving effect: $($Result[0]) - $($Result[1])"
+
+    if($Result[0] -eq "error") {
+        return @{"error" = $Result[1]; "result" = $null}
+    }
+
+    return @{"error" = $null; "result" = $Result[1]}
+}
+
+function ResolveCorrectly {
+    Param (
+        [Parameter(Mandatory=$true)] [String] $Hostname,
+        [Parameter(Mandatory=$false)] [String] $IP = "Any"
+    )
+
+    Write-Log "Trying to resolve host '$Hostname', expecting ip '$IP'"
+
+    $result = Resolve-DNS -Session $MultiNode.Sessions[0] `
+        -ContainerName $ContainersIDs[0] -Hostname $Hostname
+
+    if((-not $result.error)) {
+        if(($IP -eq "Any") -or ($result.result -eq $IP)) {
+            return $true
+        }
+    }
+    return $false
+}
+
+function ResolveWithError {
+    Param (
+        [Parameter(Mandatory=$true)] [String] $Hostname,
+        [Parameter(Mandatory=$true)] [String] $ErrorType
+    )
+
+    Write-Log "Trying to resolve host '$Hostname', expecting error '$ErrorType'"
+
+    $result = Resolve-DNS -Session $MultiNode.Sessions[0] `
+        -ContainerName $ContainersIDs[0] -Hostname $Hostname
+    return (($result.error -eq $ErrorType) -and (-not $result.result))
+}
+
+Test-WithRetries 1 {
+    Describe "DNS tests" {
+        BeforeAll {
+            Initialize-PesterLogger -OutDir $LogDir
+
+            $MultiNode = New-MultiNodeSetup -TestenvConfFile $TestenvConfFile
+            $DNSServerRepo = [DNSServerRepo]::New($MultiNode.NM)
+            [Diagnostics.CodeAnalysis.SuppressMessageAttribute(
+                "PSUseDeclaredVarsMoreThanAssignments",
+                "IPAMRepo",
+                Justification="It's actually used."
+            )]
+            $IPAMRepo = [IPAMRepo]::New($MultiNode.NM)
+
+            Write-Log "Starting Test DNS Server on second test bed..."
+            $DefaultDNSServerDir = "C:\DNS_Server"
```

Comment:
What do you think about creating here functions with calls like this `Start-TestDnsServer -Session $MultiNode.Sessions[1]...` and `Start-TestDnsResolver -Session $MultiNode.Sessions[0]...` so we could reduce the size of the BeforeAll block?

Reply:
Fixed